# ðŸ› ï¸ Generate and validate ERV gtf that matches the latest reference genome
ERV gtf annotation is needed to quantify ERV from sequencing data. But the ERV gtf compatible with the latest mouse genome GRCm39 is not publicly available. Here I generated and validated the ERV gtf based on GRCm39 using RepeatMasker and customized scripts.

> ðŸ’¡ **Tip: [Please find **ðŸ‘‰ MY BLOG** for an introduction to the project, along with the detailed instrution and annotation of the scripts in this repository.](https://myhugoblog)**

The github repo also contains excerpted introdution as described below as well as useful [untilities](util) to convert GFF or text into GTF.

## ðŸ“‘ Table of Contents
- [ðŸ“˜ Why is ERV gtf compatible with GRCm39 needed (complete version is here)](#-why-is-erv-gtf-compatible-with-grcm39-needed-complete-version-is-here)
- [ðŸ’¡ How to generate and validate the ERV gtf compatible with GRCm39](#-how-to-generate-and-validate-the-erv-gtf-compatible-with-grcm39)
  - [ðŸ’» Generate GRCm39 ERV](#-generate-grcm39-erv)
  - [ðŸ§¾ Vadlidate our GRCm39 ERV](#-vadlidate-our-grcm39-erv)
- [âœ… Conclusion](#-conclusion)

## ðŸ“˜ Why is ERV gtf compatible with GRCm39 needed (complete version is [here](myhugoblod))

GRCm39 is preferred to have most accurate and up-to-date quantification of genes or other features from RNAseq reads, because of the key improvements made in GRCm39 compared with GRCm38:
- Improved accuracy in assembly with updated sequencing data
- More corrected sequencing gaps and mis-assemblies
- Better representation of segmental duplications and structural variations
- Enhanced annotation coverage

Moreover, ERV and genes have to be quantify simultaneously for below reasons. Therefore, using the GRCm38 and its ERV annotation instead of the GRCm39 pair results in less accurate quantification of both genes and ERVs
- properly handling multi-mapping: ERV sequences can be similar or overlap with endogenous genes. If ERVs and genes arenâ€™t quantified together, reads may be incorrectly assigned to one while the other is completely ignored.
- accurately read assignment: regardless multimappings, quantifying ERV with genes ensures a good control of false positives by including a competitive pool of unmatched features.

## ðŸ’¡ How to generate and validate the ERV gtf compatible with GRCm39
[gEVE](https://pmc.ncbi.nlm.nih.gov/articles/PMC4885607/) database was generates using RepeatMasker, thus we were encouraged to generate the latest ERV gtf in the same way. We hypothesized that the ERV gtf generated by us should show a good degree of conservation when compared with GRCm38 ERV gtf in [gEVE database](http://geve.med.u-tokai.ac.jp/), and validated it correspondingly.

### ðŸ’» Generate GRCm39 ERV
1. Install RepeatMasker
2. File wrangling and run RepeatMasker
   - [faidx_split_mus38_file.sh](faidx_split_mus38_file.sh)
   - [run_repeatmasker.sh](run_repeatmasker.sh)
4. Generate GTF file from `RepeatMasker` outputs
   - please see `repmask2gtf()` in [gen_mus_annot.R](gen_mus_annot.R)
5. Modify GTF for STARsolo
   - please see `mod_gtf()` in [gen_mus_annot.R](gen_mus_annot.R)
7. Generate FASTA from GTF for validation
   - please see `mod_gtf_for_AGAT()` in [gen_mus_annot.R](gen_mus_annot.R)
   - please see [agat_gtf2fa.sh](agat_gtf2fa.sh)


### ðŸ§¾ Vadlidate our GRCm39 ERV
1. Wrange files of GRCm38 ERV
   - [faidx_split_mus38_file.sh](faidx_split_mus38_file.sh)
   - please see `mod_gtf_for_AGAT()` in [gen_mus_annot.R](gen_mus_annot.R)
   - please see [agat_gtf2fa.sh](agat_gtf2fa.sh)

3. BLASTn: gEVE as reference, our GRCm39 ERV as query
   - [gen_db.sh](gen_db.sh)
   - parrellel computation
   - [compare_fa.R](compare_fa.R)
4. BLASTn: our GRCm39 ERV as reference, gEVE as query
   - [gen_db.sh](gen_db.sh)
   - parrellel computation
   - [compare_fa_reverse.R](compare_fa_reverse.R)



## âœ… Conclusion
Our GRCm39 ERV capture nearly all information in gEVE GRCm38 ERV, and add ~4000 more unique ERV sequences. We attribute this improvement to the up-to-date GRCm39 and the sensitivity settings available in the newer version of RepeatMasker. Below observation further revealed that the newly identified sequences in GRCm39 ERV predominantly belong to ERV and LINE, both of significant interest in biomedical research.
- Out of 9043 total sequences in gEVE, 8768 has a BLASTn \>99% matched sequence in our GRCm39 ERV, corresponding to 4321 unique GRCm39 sequences
- Only 55% of GRCm39 ERV segments and 27% of GRCm39 LINE segments have matched sequences in gEVE
